feat(reports) unbalanced invoice payments report

jeremielodi opened the pull request:
This report shows the balance of invoices witch are not completely paid or overpaid during a period"

jeremielodi committed:
feat(reports) unbalanced invoice payments report

mbayopanda commented on client/src/i18n/fr/report.json:
```diff
@@ -125,6 +125,10 @@
       "DESCRIPTION" : "Ce rapport affiche tous les groupes de débiteurs, leurs soldes de compte au début de l'exercice, les soldes actuels et leurs soldes de clôture"
     },
+    "UNBALANCED_INVOICE_PAYMENTS_REPORT": {
+      "TITLE": "Rapport de la Balance des paiements non finalisés des factures",
```
A tiny comment, B of Balance can be in lowercase for having : `Rapport de la balance des paiements non finalisés des factures`

mbayopanda commented on server/models/procedures/invoicing.sql:
```diff
@@ -771,3 +771,95 @@ BEGIN
+      INSERT INTO stage_invoice_balance
+      SELECT _tempRef, _uuid, SUM(invoices.debit) AS debit, SUM(invoices.credit) AS credit, _debtor_uuid, _date, _invoiceID
+      FROM
+	      ( SELECT  p.record_uuid AS UUID, p.debit, p.credit
```
Could you use `debit_equiv` and `credit_equiv`

mbayopanda commented on server/models/procedures/invoicing.sql:
```diff
@@ -771,3 +771,95 @@ BEGIN
+      INSERT INTO stage_invoice_balance
+      SELECT _tempRef, _uuid, SUM(payments.debit) AS debit, SUM(payments.credit) AS credit, _debtor_uuid, _date, _invoiceID
+      FROM
+         ( SELECT  p.record_uuid AS UUID, p.debit, p.credit
```
Could you also use `debit_equiv` and `credit_equiv` for all data coming from `posting_journal` or `general_ledger`

mbayopanda reviewed (commented):
I made some temporary comment, at the end of the final version we can do complete review.

I noticed that in the procedure, invoice are fetched without any distinction to cancelled invoice.

mbayopanda reviewed (commented):
Is it possible to combine both insertion (invoice and payment) of stage_invoice_balance into just on insertion with the condition : 
`WHERE (p.record_uuid = _uuid AND p.entity_uuid = _debtor_uuid) OR (p.reference_uuid = _uuid AND p.entity_uuid = _debtor_uuid`)

mbayopanda commented on server/models/procedures/invoicing.sql:
```diff
@@ -771,3 +771,79 @@ BEGIN
+          SELECT p.record_uuid AS UUID, p.debit_equiv, p.credit_equiv
+          FROM general_ledger p
+          WHERE  (p.record_uuid = _uuid AND p.entity_uuid = _debtor_uuid) OR (p.reference_uuid = _uuid AND p.entity_uuid = _debtor_uuid)
+        ) AS invoices
```
Since this query is for invoice and payment, you could use a better name such as `invoices_and_payments`.

jeremielodi replied:
👍

jniles commented on client/src/i18n/en/report.json:
```diff
@@ -129,6 +129,10 @@
     },
+    "UNBALANCED_INVOICE_PAYMENTS_REPORT": {
+      "TITLE": "Unbalaced invoice payments report",
+      "DESCRIPTION": "This report shows the balance of invoices witch are not completely paid or overpaid during a period"
```
This is pretty good!  "witch" should be spelled "which"

jniles commented on client/src/modules/reports/generate/unbalanced_invoice_payments_report/unbalanced_invoice_payments_report.config.js:
```diff
@@ -0,0 +1,74 @@
+];
+
+function UnbalancedInvoicePaymentsConfigController($sce,
+  Notify, SavedReports, AppCache, reportData, $state, Moment, bhContants) {
```
Could you rename this variable `bhConstants`?  It has a slight typo.

jniles commented on server/controllers/finance/reports/unbalanced_invoice_payments/index.js:
```diff
@@ -0,0 +1,133 @@
+exports.document = build;
+
+/**
+ * Actually builds the open debtor report.
```
This isn't the open debtor report - could you either update this comment or remove it?

jniles commented on server/controllers/finance/reports/unbalanced_invoice_payments/index.js:
```diff
@@ -0,0 +1,133 @@
+async function getBalances(options) {
+
+  const { dateFrom, dateTo } = options;
+  const debtorsSql = `
```
Can't you put this SQL in the stored procedure?  That way, you won't need to loop through a large JS array?

jniles commented on server/models/bhima.sql:
```diff
@@ -141,7 +142,11 @@ INSERT INTO `report` (`id`, `report_key`, `title_key`) VALUES
   (21, 'account_reference', 'REPORT.ACCOUNT_REFERENCE.TITLE'),
+<<<<<<< 61e09ac1e4165dc23c39255468abc7b4f854ac20
   (22, 'account_report_multiple', 'REPORT.REPORT_ACCOUNTS_MULTIPLE.TITLE');
+=======
```
Looks like this is a broken merge.

jniles commented on server/models/procedures/invoicing.sql:
```diff
@@ -771,3 +771,79 @@ BEGIN
+    SUM(debit) AS debit, SUM(credit) AS credit, DATE AS creation_date, reference
+    FROM stage_invoice_balance
+    WHERE tempRef = _tempRef
+    GROUP BY reference
```
It will be faster to group by `invoice_uuid`.  It should do the same thing, I think, but `invoice_uuid` is binary.

jniles commented on server/models/procedures/invoicing.sql:
```diff
@@ -771,3 +771,79 @@ BEGIN
+  CLOSE curs1;
+
+  SELECT x.invoice_uuid, x.debtor_uuid, x.debit, x.credit, (x.debit - x.credit) AS balance,
+    (x.credit/x.debit) AS paymentPercentage , x.reference
```
If `x.debit` is 0, `paymentPercentage` is `NULL`.  Let's catch it with `IFNULL()` and make it `0`.

jniles commented on server/models/procedures/invoicing.sql:
```diff
@@ -771,3 +771,79 @@ BEGIN
+    END LOOP;
+  CLOSE curs1;
+
+  SELECT x.invoice_uuid, x.debtor_uuid, x.debit, x.credit, (x.debit - x.credit) AS balance,
```
It would be interesting to have more statistics on this.  You can easily create a lightweight temporary table to handle this case.

```sql
CREATE TEMPORARY TABLE `invoice_balances` AS
  SELECT x.invoice_uuid, x.debtor_uuid, x.debit, x.credit, (x.debit - x.credit) AS balance,
   (x.credit/x.debit) AS paymentPercentage , x.reference
  FROM (
    SELECT  BUID(invoice_uuid) AS invoice_uuid, BUID(debtor_uuid) AS debtor_uuid, 
    SUM(debit) AS debit, SUM(credit) AS credit, DATE AS creation_date, reference
    FROM stage_invoice_balance
    WHERE tempRef = _tempRef
    GROUP BY reference
  ) AS x
  WHERE  x.debit <> x.credit;
```

Now you can do things like this:
```sql
SET totalInvoiceValue = (SELECT SUM(debit) FROM invoice_balances);
SELECT
  SUM(IF(balance / totalInvoiceValue < 0.25, 1, 0) as _0_25,
  SUM(IF(balance / totalInvoiceValue >= 0.25 AND (balance / totalInvoiceValue) < 0.50, 1, 0) AS _25_50,
  SUM(IF(balance / totalInvoiceValue >= 0.50 AND (balance / totalInvoiceValue) < 0.75, 1, 0) AS _50_75,
  SUM(IF(balance / totalInvoiceValue >= 0.75 AND (balance / totalInvoiceValue) < 1, 1, 0) AS _75_100,
  SUM(IF(balance / totalInvoiceValue > 100, 1, 0) AS 100_plus,
FROM invoice_balances;
```
to generate any statistics you need.

jniles commented on server/models/procedures/invoicing.sql:
```diff
@@ -771,3 +771,79 @@ BEGIN
+  SELECT x.invoice_uuid, x.debtor_uuid, x.debit, x.credit, (x.debit - x.credit) AS balance,
+    (x.credit/x.debit) AS paymentPercentage , x.reference
+  FROM (
+    SELECT  BUID(invoice_uuid) AS invoice_uuid, BUID(debtor_uuid) AS debtor_uuid, 
```
What happens if you change this query to..
```sql
    SELECT  BUID(invoice_uuid) AS invoice_uuid, BUID(debtor_uuid) AS debtor_uuid,
      SUM(debit) AS s_debit, SUM(credit) AS s_credit, date AS creation_date, reference
   FROM stage_invoice_balance
   WHERE tempRef = _tempRef
   GROUP BY reference
   HAVING s_debit <> s_credit;
```
?  Does it produce the same results (except for the variables renamed)?

jniles replied:
Also, why do you need `BUID()`?

jniles commented on server/models/procedures/invoicing.sql:
```diff
@@ -771,3 +771,79 @@ BEGIN
+
+  -- all invoice balances will be stored temporarely in stage_invoice_balance table
+
+  CREATE TEMPORARY TABLE IF NOT EXISTS stage_invoice_balance(
```
You should definitely move to check to outside of the LOOP.

jniles commented on server/models/procedures/invoicing.sql:
```diff
@@ -771,3 +771,79 @@ BEGIN
+          UNION ALL
+          SELECT p.record_uuid AS UUID, p.debit_equiv, p.credit_equiv
+          FROM general_ledger p
+          WHERE  (p.record_uuid = _uuid AND p.entity_uuid = _debtor_uuid) OR (p.reference_uuid = _uuid AND p.entity_uuid = _debtor_uuid)
```
In my experience, `OR` in SQL is super inefficient.   It would be better to break these into two different conditions and then put them together with `UNION ALL`.

jniles commented on server/models/procedures/invoicing.sql:
```diff
@@ -771,3 +771,79 @@ BEGIN
+      INSERT INTO stage_invoice_balance
+      SELECT _tempRef, _uuid, SUM(invoices_payments.debit_equiv) AS debit, SUM(invoices_payments.credit_equiv) AS credit, _debtor_uuid, _date, _invoiceID
+      FROM
+	      ( SELECT  p.record_uuid AS UUID, p.debit_equiv, p.credit_equiv
```
MySQL is not case-sensitive on all systems (see [here](https://dev.mysql.com/doc/refman/5.7/en/identifier-case-sensitivity.html)), so it is better to make every variable lowercase in MySQL.

jniles commented on server/models/procedures/invoicing.sql:
```diff
@@ -771,3 +771,79 @@ BEGIN
+
+  DECLARE done BOOLEAN;
+  DECLARE curs1 CURSOR FOR 
+    SELECT  i.uuid, i.debtor_uuid, i.date 
```
Would it be possible to just create a temporary table to hold invoice uuids?  A JOIN might be faster than a LOOP since it can use indexes more efficiently.  Like this:
```sql
CREATE TEMPORARY TABLE `invoices_to_find` AS 
  SELECT invoice.uuid, debtor_uuid, dm.text AS reference
  FROM invoice JOIN document_map dm ON invoice.uuid = dm.uuid
  WHERE DATE(i.date) BETWEEN DATE(dateFrom) AND DATE(dateTo);
```

Then we can do something like this: (I've renamed `UUID` to `i_uuid`, to make things clearer)
```sql
INSERT INTO stage_invoice_balance
  SELECT i_uuid, SUM(invoices_payments.debit_equiv) AS debit, SUM(invoices_payments.credit_equiv) AS credit
  FROM (

    -- posting journal invoices
    SELECT p.record_uuid AS i_uuid, p.debit_equiv, p.credit_equiv
      FROM posting_journal p JOIN invoices_to_find iv ON p.record_uuid = iv.uuid
      WHERE iv.debtor_uuid = p.entity_uuid

    UNION ALL

    -- general ledger invoices
    SELECT p.record_uuid AS i_uuid, p.debit_equiv, p.credit_equiv
      FROM general_ledger p JOIN invoices_to_find iv ON p.record_uuid = iv.uuid
      WHERE iv.debtor_uuid = p.entity_uuid

    UNION ALL

    -- posting_journal payments
    SELECT p.reference_uuid AS i_uuid, p.debit_equiv, p.credit_equiv
      FROM posting_journal p JOIN invoices_to_find iv ON p.reference_uuid = iv.uuid

    UNION ALL

    -- general_ledger payments
    SELECT p.reference_uuid AS i_uuid, p.debit_equiv, p.credit_equiv
      FROM general_ledger p JOIN invoices_to_find iv ON p.reference_uuid = iv.uuid
  )z
  GROUP BY i_uuid;
```

jeremielodi committed:
migration info

jniles committed:
perf: rewrite UnbalancedInvoicePayments() SP

This commit rewrites the UnbalanedInvoicePayments stored procedure to
leveraged the INDEX on reference_uuid.  The SP uses temporary tables to
group invoices and payments against them, before combining each to find
their balance.

jniles committed:
Merge remote-tracking branch 'upstream/master' into invoice_payment_balance

jniles committed:
perf(reports): reduce number of JS loops

This commit reduces the number of JS loops by using lodash.  It is also
semantically cleaner with commented code.

jniles committed:
chore(reports): clean up report presentation

This commit cleans up the report presentation by moving the summary
tables to be side by side.

jniles committed:
Merge remote-tracking branch 'upstream/master' into HEAD
